File: rtsp/src/main/java/com/pedro/rtsp/rtp/packets/BasePacket.java
package com.pedro.rtsp.rtp.packets;

import android.media.MediaCodec;
import com.pedro.rtsp.utils.RtpConstants;
import java.nio.ByteBuffer;
import java.util.Random;

/**
 * Created by pedro on 27/11/18.
 */

public abstract class BasePacket {

  protected final static int maxPacketSize = RtpConstants.MTU - 28;
  protected byte channelIdentifier;
  protected int rtpPort;
  protected int rtcpPort;
  private final long clock;
  private int seq = 0;
  private int ssrc;

  public BasePacket(long clock) {
    this.clock = clock;
    ssrc = new Random().nextInt();
  }

  public abstract void createAndSendPacket(ByteBuffer byteBuffer, MediaCodec.BufferInfo bufferInfo);

  public void setPorts(int rtpPort, int rtcpPort) {
    this.rtpPort = rtpPort;
    this.rtcpPort = rtcpPort;
  }

  public void reset() {
    seq = 0;
    ssrc = new Random().nextInt();
  }

  protected byte[] getBuffer(int size) {
    byte[] buffer = new byte[size];
    buffer[0] = (byte) Integer.parseInt("10000000", 2);
    buffer[1] = (byte) RtpConstants.payloadType;
    setLongSSRC(buffer, ssrc);
    requestBuffer(buffer);
    return buffer;
  }

  protected void updateTimeStamp(byte[] buffer, long timestamp) {
    long ts = timestamp * clock / 1000000000L;
    setLong(buffer, ts, 4, 8);
  }

  protected void setLong(byte[] buffer, long n, int begin, int end) {
    for (end--; end >= begin; end--) {
      buffer[end] = (byte) (n % 256);
      n >>= 8;
    }
  }

  protected void updateSeq(byte[] buffer) {
    setLong(buffer, ++seq, 2, 4);
  }

  protected void markPacket(byte[] buffer) {
    buffer[1] |= 0x80;
  }

  private void setLongSSRC(byte[] buffer, int ssrc) {
    setLong(buffer, ssrc, 8, 12);
  }

  private void requestBuffer(byte[] buffer) {
    buffer[1] &= 0x7F;
  }
}

File: rtsp/src/main/java/com/pedro/rtsp/rtp/packets/BasePacket.kt
/*
 * Copyright (C) 2024 pedroSG94.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pedro.rtsp.rtp.packets

import com.pedro.common.frame.MediaFrame
import com.pedro.rtsp.rtsp.RtpFrame
import com.pedro.rtsp.utils.RtpConstants
import com.pedro.rtsp.utils.setLong
import kotlin.experimental.and
import kotlin.experimental.or

/**
 * Created by pedro on 27/11/18.
 */
abstract class BasePacket(private var clock: Long, private val payloadType: Int) {

  protected var channelIdentifier: Int = 0
  private var seq = 0L
  private var ssrc = 0L
  protected val maxPacketSize = RtpConstants.MTU - 28
  protected val TAG = "BasePacket"

  abstract suspend fun createAndSendPacket(
    mediaFrame: MediaFrame,
    callback: suspend (List<RtpFrame>) -> Unit
  )

  open fun reset() {
    seq = 0
    ssrc = 0
  }

  fun setSSRC(ssrc: Long) {
    this.ssrc = ssrc
  }

  protected fun setClock(clock: Long) {
    this.clock = clock
  }

  protected fun getBuffer(size: Int): ByteArray {
    val buffer = ByteArray(size)
    buffer[0] = 0x80.toByte()
    buffer[1] = payloadType.toByte()
    setLongSSRC(buffer, ssrc)
    requestBuffer(buffer)
    return buffer
  }

  protected fun updateTimeStamp(buffer: ByteArray, timestamp: Long): Long {
    val ts = timestamp * clock / 1000000000L
    buffer.setLong(ts, 4, 8)
    return ts
  }

  protected fun updateSeq(buffer: ByteArray) {
    buffer.setLong(++seq, 2, 4)
  }

  protected fun markPacket(buffer: ByteArray) {
    buffer[1] = buffer[1] or 0x80.toByte()
  }

  private fun setLongSSRC(buffer: ByteArray, ssrc: Long) {
    buffer.setLong(ssrc, 8, 12)
  }

  private fun requestBuffer(buffer: ByteArray) {
    buffer[1] = buffer[1] and 0x7F
  }
}